### DISCUSSION TITLE
Trouble with LRO NAC Stereo

### DISCUSSION THREAD

**Discussion by FlashRisk** [NeoGeographyToolkit/StereoPipeline#408] [Q&A] [answered]
I am currently having trouble getting a stereo image from LRO NAC images. I am downloading NAC images and JSON Labels from PILOT, then I am running the following code for each respective image: 
f=file name(ex. M1112693469LE)
lronac2isis from = ${f}.IMG     to = ${f}.cub
spiceinit   from = ${f}.cub shape = ellipsoid web = true
lronaccal   from = ${f}.cub     to = ${f}.cal.cub
lronacecho  from = ${f}.cal.cub to = ${f}.cal.echo.cub

(^These commands execute without a hitch.)

I then crop each cal.echo.cub file with qview. Finally, I run one of the 3 following code options:
1. 
parallel_stereo M1114696273RE.cal.echo.cub  \
  M1219484122LE.cal.echo.cub result/output  \
  --alignment-method affineepipolar

2. 
stereo_gui M1106796038RE_crop.cal.echo.cub M1112693469LE_crop.cal.echo.cub

3. 
stereo_gui M1106796038RE_crop.cal.echo.cub M1112693469LE_crop.cal.echo.cub \
  M1106796038RE.json M1112693469LE.json                  \
  --alignment-method local_epipolar                    \
 --stereo-algorithm asp_mgm --subpixel-mode 9         \
  run/run

When I run the first option I get the following error: 
/home/usr/miniconda3/envs/asp/bin/stereo_parse: symbol lookup error: /home/usr/miniconda3/envs/asp/bin/../lib/libisis.so: undefined symbol: _ZN3ale4loadENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES5_S5_b
	--> Setting number of processing threads to: 4
Stereo file ./stereo.default could not be found. Will use default settings and command line options only.
num_stereo_pairs,1
Using session: isis
Loading camera model: M1106796038RE_crop.cal.echo.cub 

None
Traceback (most recent call last):
  File "/home/usr/miniconda3/envs/asp/bin/parallel_stereo", line 897, in <module>
    settings = run_and_parse_output("stereo_parse", args, sep, opt.verbose)
  File "/home/usr/miniconda3/envs/asp/libexec/asp_system_utils.py", line 294, in run_and_parse_output
    raise Exception('Failed executing: ' + " ".join(call))
Exception: Failed executing: /home/usr/miniconda3/envs/asp/bin/stereo_parse M1106796038RE_crop.cal.echo.cub M1112693469LE_crop.cal.echo.cub result/output --alignment-method affineepipolar --corr-seed-mode 1

When I try to run stereo_parallel on the GUI I get the message: "No valid region for stereo is present. Regions can be selected with Control-Mouse in each image." I can see that the region is the same, except one is upside down. Does the orientation of the image make a difference?

Any help would be greatly appreciated. Thanks!

**Comment by oleg-alexandrov** [comment]
>
>
> > stereo_gui M1106796038RE_crop.cal.echo.cub
> M1112693469LE_crop.cal.echo.cub
>
Here there is no output prefix, as the third option, if you want to run
stereo from inside the gui, but your issues are likely not here.


> /home/usr/miniconda3/envs/asp/bin/stereo_parse: symbol lookup error:
> /home/usr/miniconda3/envs/asp/bin/../lib/libisis.so: undefined symbol:
> _ZN3ale4loadENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES5_S5_b
>
>
I never saw this before. Your "libale" is missing. Normally conda should
fetch that. You can try the binary release (non-conda) from:
https://stereopipeline.readthedocs.io/en/latest/installation.html


> When I try to run stereo_parallel on the GUI I get the message: "No valid
> region for stereo is present. Regions can be selected with Control-Mouse in
> each image." I can see that the region is the same, except one is upside
> down. Does the orientation of the image make a difference?
>

No. Orientation does not make a difference. Did you select the same
visually similar images with Control-Mouse? It should draw a rectangle
around each as in the screenshot in the manual.

Let us know how it goes.

**Reply by FlashRisk** [reply]
I reinstalled ASP with the binary release, but I am still encountering some errors when running stereo. I also looked at the manual and I selected two sections that were as similar as possible. When running the stereo_gui command it now returns: "Done running stereo." But nothing appears in the run or result file. This is what appears in the terminal:

$ stereo_gui M1106796038RE_crop.cal.echo.cub M1112693469LE_crop.cal.echo.cub \
>   M1106796038RE.json M1112693469LE.json                  \
>   --alignment-method local_epipolar                    \
>  --stereo-algorithm asp_mgm --subpixel-mode 2        \
>   run/run
	--> Setting number of processing threads to: 4
Not a valid image: run/run. Using this as the output prefix.
Reading: M1106796038RE_crop.cal.echo.cub
Detected large image: M1106796038RE_crop.cal.echo.cub.
Will construct an image pyramid on disk.
Using existing subsampled image: M1106796038RE_crop.cal.echo_sub2.tif
Using existing subsampled image: M1106796038RE_crop.cal.echo_sub4.tif
Using existing subsampled image: M1106796038RE_crop.cal.echo_sub8.tif
Reading: M1112693469LE_crop.cal.echo.cub
Detected large image: M1112693469LE_crop.cal.echo.cub.
Will construct an image pyramid on disk.
Using existing subsampled image: M1112693469LE_crop.cal.echo_sub2.tif
Using existing subsampled image: M1112693469LE_crop.cal.echo_sub4.tif
Using existing subsampled image: M1112693469LE_crop.cal.echo_sub8.tif
qt.qpa.xcb: QXcbConnection: XCB error: 14 (BadIDChoice), sequence: 380, resource id: 1, major code: 53 (CreatePixmap), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 9 (BadDrawable), sequence: 381, resource id: 1, major code: 55 (CreateGC), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 9 (BadDrawable), sequence: 382, resource id: 1, major code: 72 (PutImage), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 13 (BadGC), sequence: 383, resource id: 71303177, major code: 60 (FreeGC), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 9 (BadDrawable), sequence: 384, resource id: 1, major code: 139 (Unknown), minor code: 4
qt.qpa.xcb: QXcbConnection: XCB error: 4 (BadPixmap), sequence: 385, resource id: 1, major code: 54 (FreePixmap), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 143 (Unknown), sequence: 386, resource id: 71303178, major code: 139 (Unknown), minor code: 27
qt.qpa.xcb: QXcbConnection: XCB error: 143 (Unknown), sequence: 387, resource id: 71303178, major code: 139 (Unknown), minor code: 7
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 390, resource id: 71303179, major code: 138 (Unknown), minor code: 23
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 391, resource id: 71303179, major code: 138 (Unknown), minor code: 23
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 392, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 393, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
src win for    M1106796038RE_crop.cal.echo.cub: 562 945 3174 2868
src win for    M1112693469LE_crop.cal.echo.cub: 362 743 3265 3095
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 7591, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 7592, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
Running: /home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/parallel_stereo M1106796038RE_crop.cal.echo.cub M1112693469LE_crop.cal.echo.cub M1106796038RE.json M1112693469LE.json --alignment-method local_epipolar --stereo-algorithm asp_mgm --subpixel-mode 2 run/run --left-image-crop-win 562 945 3174 2868 --right-image-crop-win 362 743 3265 3095
	--> Setting number of processing threads to: 4
Stereo file ./stereo.default could not be found. Will use default settings and command line options only.
num_stereo_pairs,1
Will refine the disparity using the ASP subpixel-mode: 2.
Using session: csm
Loading camera model: M1106796038RE_crop.cal.echo.cub M1106796038RE.json
Unknown unit for the ellipsoid radii in M1106796038RE.json. The read value is:
Traceback (most recent call last):
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/parallel_stereo", line 898, in <module>
    settings = run_and_parse_output("stereo_parse", args, sep, opt.verbose)
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/asp_system_utils.py", line 298, in run_and_parse_output
    raise Exception('Failed executing: ' + " ".join(call))
Exception: Failed executing: /home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/stereo_parse M1106796038RE_crop.cal.echo.cub M1112693469LE_crop.cal.echo.cub M1106796038RE.json M1112693469LE.json --alignment-method local_epipolar --stereo-algorithm asp_mgm --subpixel-mode 2 run/run --left-image-crop-win 562 945 3174 2868 --right-image-crop-win 362 743 3265 3095 --corr-seed-mode 1
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 7789, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 7790, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
$

Thank you so much for the reply, I hope this isn't too much of a hastle.

**Comment by oleg-alexandrov** [comment]
This is actually progress. Sorry for the gui noise, that is some Qt
thingie.

Now you are getting the error: Unknown unit for the ellipsoid radii in
M1106796038RE.json.

I will suggest you open up this .json file. I see in my own .json file text
like:

"radii": {
    "semimajor": 1737.4,
    "semiminor": 1737.4,
    "unit": "km"
  },

I wonder what you see. Apparently we do not allow any other unit except
meters and km.

This is the first time I see something else in there. I think the unit is
important, otherwise we cannot convert properly to meters.

You can also edit those files and convert the values and units to meters,
and that should let it run, I think.

**Reply by FlashRisk** [reply]
So turns out there is not a variable that matches the ellipsoid radii name. The closest thing that I found was the following code:

"naifkeywords": {
        "body_code": 301,
        "body301_radii": [
          1737.4,
          1737.4,
          1737.4
],

I wonder if it has something to do with downloading the .json file from PILOT? Should I try downloading the .json file from somewhere else?

Also thank you for replying so fast and offering such good advice. It's really appreciated.

**Comment by oleg-alexandrov** [comment]
Aha. I don't know what "PILOT" is, and I could not find any info online.
Maybe one needs to figure out what tools they use to create .json files.
ASP uses, what I thought until now, the ale library (
https://github.com/DOI-USGS/ale).

We have an example for LRO NAC at
https://stereopipeline.readthedocs.io/en/latest/examples/csm.html (section
on linescan cameras, with prior material there for context).

**Comment by dpmayerUSGS** [comment]
@FlashRisk The JSON labels on [PILOT](https://pilot.usgs.gov) are provided as a convenience for users who might want to do advanced searching through PDS labels, but the JSON files themselves are not part of the PDS archive and aren't meant to be ingested by applications like ASP. Don't pass the JSON files to ASP. You don't need the information from the JSON files for ASP anyway because the labels are already embedded in the ISIS cubes that you've prepared. 

Also, ASP has a script that automates preprocessing of LROC NAC stereopairs, [`lronac2mosaic.py`](https://stereopipeline.readthedocs.io/en/latest/tools/lronac2mosaic.html). Have you taken a look at that?

**Comment by rbeyer** [comment]
So the confusion here is that JSON is just a way to organize information in a text file, like CSV or YAML.

As @dpmayerUSGS points out, the PILOT files are to support searching, but the JSON files that you can give to ASP must contain Community Sensor Model information, they coincidentally are both JSON files, but have wildly different contents.

The documentation Oleg points to describes how to create `gen_csm.py` which will produces the CSM JSON files that ASP can use.  Modern versions of ALE install an `isd_generate` program into the environment. If your ISIS or ASP conda environment has ALE (check via `conda list`) then you can see if you can run `isd_generate -h` at your prompt (without having to copy and paste to create `gen_csm.py`).  It is meant to do the same thing that the ASP documentation uses `gen_csm.py` for.

**Comment by FlashRisk** [comment]
Thank you @rbeyer @dpmayerUSGS and @oleg-alexandrov for the advice, it really helps. I haven't worked with .json files before and assumed they were a part of the PDS archive. However when using the both the stereo_gui and the parallel_stereo commands with just the .cub files I get the following error:

Error: **ERROR** Unable to initialize camera model in Camera Factory.
**I/O ERROR** Spice file does not exist [/home/usr/miniconda3/envs/isis/lro/kernels/pck/moon_080317.tf].

I don't know if this makes any difference, but for the preprocessing of the images I ran the following code:

f=M1112693469LE
lronac2isis from = ${f}.IMG     to = ${f}.cub
spiceinit   from = ${f}.cub shape = ellipsoid web = true
lronaccal   from = ${f}.cub     to = ${f}.cal.cub

I don't have local SPICE data and instead am using the ISIS SPICE Web Service as I don't have enough space on my computer for all of the kernels. I do have an external hard drive that I might be able to hook up if need be...

Also, I haven't worked with the terminal, python, or C++ prior to this so I am a beginner when it comes to all this. If need be I can produce the CSM JSON files using gen_csm.py, but if there is an easier way to do it I would prefer to do that.

 Sorry if these are stupid questions.

**Comment by oleg-alexandrov** [comment]
I think what you may be seeing here is that while spiceinit can get away
with using web-based kernels, for creating the .cub file, that is not
enough for creating .json files. That needs local kernels (also setting
ISISDATA, ALESPICEROOT).

The lro data kernels seem to take up 270 GB on my hard drive, and that is
humongous.

The trick I use sometimes is to simply download all the kernels it is
complaining about, such as what you see earlier, with lro/kernels/pck/
moon_080317.tf. A list of these should be in the .cub file, I think, though
it may not be complete. So, several attempts may be needed, but it beats
waiting half a day to get the full set.

Also note the remark here:
https://stereopipeline.readthedocs.io/en/latest/examples/csm.html#creation-of-csm-cameras.
That if you get an error about your paths not being the same as the
hard-coded path /usgs/cpkgs/isis3/.

Sorry for all this pain. ISIS is a legacy system, and while it is also
being modernized, there's more to go.

**Comment by FlashRisk** [comment]
I was able to make some progress! Although I encountered a new error. The following is a portion of what the program returned:

Performing IP matching with alignment.
        Rough homography--> [****************************************] Complete!
Projected 18184 rays for rough homography.
Number of inliers: 18184.
Aligning right to left for IP capture using rough homography: Matrix3x3((-0.911702,-0.0555786,0)(0.0624794,-0.904369,0)(-7.44586e-07,-2.15614e-08,1))
	    Looking for IP in left image...
	    Using 911 interest points per tile (1024^2 px).
	    Detecting IP
	    Removing IP near nodata with radius 4
	    Building descriptors
	    Found interest points: 5002
	    Recording interest points to file: run/run-L-cropped.vwip
	    Looking for IP in right image...
	    Using 944 interest points per tile (1024^2 px).
	    Detecting IP
	    Removing IP near nodata with radius 4
	    Building descriptors
	    Found interest points: 4415
	--> Matching interest points using the epipolar line.
	    Uniqueness threshold: 0.8
	    Epipolar threshold:   3497.93
	    Matching forward
	    ---> Obtained 5002 matches.
	    Matching backward
	    ---> Obtained 4415 matches.
	    Matched 21 points.
	    Inlier cluster:
	      Triangulation error: 497.697 +- 473.934 meters
	      Removed 6 points in triangulation filtering.
	      Removed 0 points in stddev filtering.
	    Reduced matches to 15
	    * Writing match file: run/run-L-cropped__R-cropped.match
RANSAC was unable to find a fit that matched the supplied data.


VW Error: RANSAC was unable to find a fit that matched the supplied data.
Traceback (most recent call last):
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/parallel_stereo", line 1090, in <module>
    normal_run('stereo_pprc', args, msg='%d: Preprocessing' % step)
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/parallel_stereo", line 679, in normal_run
    raise Exception('Stereo step ' + kw['msg'] + ' failed')
Exception: Stereo step 0: Preprocessing failed
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 1846, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
qt.qpa.xcb: QXcbConnection: XCB error: 6 (BadCursor), sequence: 1847, resource id: 71303179, major code: 2 (ChangeWindowAttributes), minor code: 0
$

Edit: I ran it a few more times with the error above happening a few more times. But occasionally the following error will occur:

[ 2023-Aug-07 14:37:23 ] : Stage 1 --> LOW-RESOLUTION CORRELATION
Cached IP match file found: run/run-L-cropped__R-cropped.match
	--> Using interest points to determine search window.
	    * Loading match file: run/run-L-cropped__R-cropped.match
Removed 1 outliers based on percentiles of differences of interest points with --outlier-removal-params.
Number of IPs left after filtering is 17 which is less than the required amount of 30, aborting stereo_corr. A solution may be to remove the run directory and restart stereo, while setting --ip-per-tile 200 or some other larger number. Otherwise decrease --min-num-ip to accept these matches. Also check your input images.

Traceback (most recent call last):
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/parallel_stereo", line 1104, in <module>
    calc_lowres_disp(args, opt, sep, resume = opt.resume_at_corr)
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/stereo_utils.py", line 273, in calc_lowres_disp
    stereo_run('stereo_corr', local_args, opt, msg='')
  File "/home/usr/Downloads/StereoPipeline-3.2.1-alpha-2023-08-04-x86_64-Linux/libexec/stereo_utils.py", line 197, in stereo_run
    raise Exception('Stereo step ' + kw['msg'] + ' failed')
Exception: Stereo step  failed


(P.S. Sorry for all of the questions I am asking. I hope this isn't too much of a hastle.)

**Comment by oleg-alexandrov** [comment, marked as answer]
Glad to hear you made progress.

The error:

> Matching forward
> ---> Obtained 5002 matches.
> Matching backward
> ---> Obtained 4415 matches.
> Matched 21 points.
> Inlier cluster:
> Triangulation error: 497.697 +- 473.934 meters
> Removed 6 points in triangulation filtering.
> Removed 0 points in stddev filtering.
> Reduced matches to 15
> * Writing match file: run/run-L-cropped__R-cropped.match
> RANSAC was unable to find a fit that matched the supplied data.

is  sign of the left and right images not being similar, as most interest
points failed to match. I will suggest you examine your images side by
side, and seeing if they are similar, if the shadows are in the same place,
etc.

Sometimes LRO NAC images are acquired with vastly different illumination,
and then it will fail.

If that's the case, you may need to fetch new data.

Another thing to watch for is convergence angle.

Here's a writeup of these issues:
https://stereopipeline.readthedocs.io/en/latest/examples/stereo_pairs.html

**Comment by oleg-alexandrov** [comment]
To add to the previous, if you are sure the images are very similar, you
can also consider selecting bigger areas in both clips, if you are using
the gui (ast it seems based on your log).

**Comment by FlashRisk** [comment]
I was able to get a result! I used different data and that seemed to do the trick. Thank you so much @oleg-alexandrov. You were so helpful through all of my blunders. I will toy around with the different stereo algorithms and other parameters to get some improved results.

**Comment by oleg-alexandrov** [comment]
Glad to hear there is progress. 

I made the error message more informative, by stating, before these errors, that the user should check the images for similarity in illumination and overlap. This will be soon in the checked in code and for a future release.
